### ISSUE TITLE
Support mainstream pruning techniques 

### ISSUE THREAD

**Issue by he-y** [pytorch/pytorch#39765]
## 🚀 Feature
support mainstream pruning techniques.

## Motivation
Recently, lots of new pruning algorithms are proposed, but the [current implementation](https://github.com/pytorch/pytorch/blob/4fef3763dd7266195e8ef20c0a5d4dd1219afeb0/torch/nn/utils/prune.py) mainly contains norm-based pruning methods. 

Based on the [discussion here](https://github.com/pytorch/pytorch/issues/38598#issuecomment-641009728), I believe it's meaningful to add some state-of-the-art pruning methods.

**Comment by he-y** [comment]
The pruning methods could be roughly divided into two categories, structured pruning, and unstructured pruning.
I mainly list a few **structured** pruning methods based on chronology order here, as I know little about **unstructured** pruning.

- Structured Pruning
1. ICLR'17, ‎Cited by 1059.
Pruning Filters for Efficient ConvNets
2. ICCV'17, ‎Cited by 690.
Channel pruning for accelerating very deep neural networks
3. ICCV'17, ‎Cited by 516.
ThiNet: A Filter Level Pruning Method for Deep Neural Network Compression
4. ICCV'17, ‎Cited by 495.
Learning Efficient Convolutional Networks Through Network Slimming
5. ICLR'17, Cited by 428.
Pruning Convolutional Neural Networks for Resource Efficient Inference
6. CVPR'18, Cited by 428.
NISP: Pruning Networks using Neuron Importance Score Propagation
7. IJCAI'18 Cited by 183.
Soft Filter Pruning for Accelerating Deep Convolutional Neural Networks
8. ICLR'19, Cited by 107.
The Lottery Ticket Hypothesis: Finding Sparse, Trainable Neural Networks
9. NIPS'19,  ‎Cited by 117.
Discrimination-aware Channel Pruning for Deep Neural Networks
10. CVPR'19, Cited by 89
Filter Pruning via Geometric Median for Deep Convolutional Neural Networks Acceleration


- Unstructured Pruning
1. ICLR'16, Cited by 3531.
Deep Compression: Compressing Deep Neural Networks with Pruning, Trained Quantization and Huffman Coding.

**Comment by mickypaganini** [comment]
@he-y thanks for providing this list.
I'm going through it and I'll be posting thoughts as I go along.

1. I believe _Pruning Filters for Efficient ConvNets_ is already implemented. You can prune entire filters (as opposed to weights) by their L1-norm using `prune.ln_structured` with the correct value of `dim`, indicating whether we want to remove input (`dim=1`) or output (`dim=0`) feature maps. The only difference with what the paper does in the procedure in Sec. 3.1 is that we don't also explicitly and automatically remove the downstream kernels that process a now removed feature map (point 3 in their procedure).

**Comment by jiecaoyu** [comment]
Following @he-y 's comments:

* Structured pruning:
Besides filter/channel-level pruning, using fine-grained block sparsity (e.g., 1x4, 1x16, 4x4) sometimes benefit more for CPU (x86, Arm) inference. Therefore, SIMD-aware weight pruning proposed in "Scalpel: Customizing DNN Pruning to the Underlying Hardware Parallelism" (ISCA'17, cited by 180) might be helpful.

* Unstructured pruning:
ADMM ("A Systematic DNN Weight Pruning Framework using Alternating Direction Method of Multipliers", ECCV'18, cited by 80) is one of the SOTA techniques, and I have validated good pruning results on various vision models.

**Comment by he-y** [comment]
@mickypaganini 
Thanks for the reply.
Yes. I agree that _Pruning Filters for Efficient ConvNets_ is already implemented via `prune.ln_structured`.
Maybe add this paper as a reference to [pruning tutorial], or comments in `prune.ln_structured` could provide extra information for readers.
Do you have any idea about the implementation of other methods?

@jiecaoyu 
Thanks for your recommendation. Do you have any idea about the implementation of these methods into [Pytorch pruning module](https://github.com/pytorch/pytorch/issues/20402).

**Comment by jiecaoyu** [comment]
Hi @he-y , I have some thoughts on the implementation:

- For the fine-grained block sparsity, ```l1_unstructured``` can be directly used with some modifications. When [computing the mask](https://github.com/pytorch/pytorch/blob/4fef3763dd7266195e8ef20c0a5d4dd1219afeb0/torch/nn/utils/prune.py#L510), we can put weights in blocks with the same size (e.g., 4), and use the square sum of the weights in each block to measure the importance of weight blocks. This is a simple example:
```python
block_size = 4

if nparams_toprune != 0:
    # reshape the size of t from [N,C,H,W] to [N,CxHxW]
    t_reshape = t.view(t.shape[0], -1)
    assert(t.shape[1] % block_size == 0) # here only support t.shape[1] % block_size == 0
    t_reshape = t_reshape.view(t_reshape.shape[0], -1, block_size)
    # calcualte the mean square of each weight block to measure the importance
    t_block_importance = t_reshape.pow(2.0).sum(2)
    assert(nparams_toprune > block_size)
    # pick the threshold
    threshold = torch.topk(
        t_block_importance.view(-1), k=(nparams_toprune // block_size), largest=False
    ).values[-1]
    t_block_mask = t_block_importance.le(threshold)
    # transfer the block mask into the mask for individual weights
    t_mask = t_block_mask.view(t_block_mask.shape[0], t_block_mask.shape[1], 1)\
        .repeat(1, 1, block_size)\
        .view(t.shape)
    mask[t_mask] = 0
```
However, in real implementation, this part can be very complicated because we need to support different weight mapping (e.g., NCHW, NHWC, NC4HW4) and different block shapes.

* Unstructured ADMM pruning:
The necessary operations can be implemented using hooks (i.e., ```register_backward_hook()```).

**Comment by ArmandXiao** [comment]
Hi @albanD @mickypaganini @ezyang,

The structured pruning method [FPGM (Filter Pruning via Geometric Median)](https://arxiv.org/abs/1811.00250) is implemented #95689 .

Please help to review the functions and usage, thank you.

---

A simple test code 👇 
```py
if __name__ == '__main__':
    import torch.nn as nn
    module = nn.Conv2d(3,20,3)
    shape = module.weight.shape
    updated_module = FPGM_structured(module, 'weight', amount=0.4, dist_type="l2", dim=0)
    # print(updated_module.weight)
    print(updated_module.state_dict().keys())

    remaining = torch.count_nonzero(torch.tensor(updated_module.weight)).item()
    total = shape[0] * shape[1] * shape[2] * shape[3]
    print("total: ", total)
    print("remaining: ", remaining)
    print("Percentage pruend:", (total - remaining) / total * 100)
```

**Comment by jbschlosser** [comment]
@supriyar do you know who I can ping on the AO side to look into this?

**Comment by supriyar** [comment]
> @supriyar do you know who I can ping on the AO side to look into this?

@jcaip can you help with this?

**Comment by jcaip** [comment]
Sure, @jbschlosser are you looking for someone to review the most recent PR?

**Comment by ArmandXiao** [comment]
Hi @albanD @jbschlosser @supriyar @jcaip , can you help to review the PR #95689 . Many thanks.

> Hi @albanD @mickypaganini @ezyang,
> 
> The structured pruning method [FPGM (Filter Pruning via Geometric Median)](https://arxiv.org/abs/1811.00250) is implemented #95689 .
> 
> Please help to review the functions and usage, thank you.
> 
> A simple test code 👇
> 
> ```python
> if __name__ == '__main__':
>     import torch.nn as nn
>     module = nn.Conv2d(3,20,3)
>     shape = module.weight.shape
>     updated_module = FPGM_structured(module, 'weight', amount=0.4, dist_type="l2", dim=0)
>     # print(updated_module.weight)
>     print(updated_module.state_dict().keys())
> 
>     remaining = torch.count_nonzero(torch.tensor(updated_module.weight)).item()
>     total = shape[0] * shape[1] * shape[2] * shape[3]
>     print("total: ", total)
>     print("remaining: ", remaining)
>     print("Percentage pruend:", (total - remaining) / total * 100)
> ```
